### PR TITLE
raidboss: replace GameLog lines (SB)

### PR DESCRIPTION
**Pull request by wexxlee**
Wasn't able to do much with the eureka stuff because we just don't have the log lines to replace the `GameLog` triggers.  See comments, particularly in `eureka_hydatos.ts`.

### CODE REVIEW

File: ui/raidboss/data/04-sb/eureka/eureka_anemos.ts [hunk 1]
```diff
@@ -179,6 +179,9 @@ const triggerSet: TriggerSet<Data> = {
       netRegex: { name: 'Pazuzu', capture: false },
       run: (data) => data.wraithCount = 0,
     },
+    // https://xivapi.com/LogMessage/916
+    // en: 7 minutes have elapsed since your last activity. [...]
+    // But no 0x29 log line; so have to use GameLog for now.
```
Comment (valarnin) [inline ui/raidboss/data/04-sb/eureka/eureka_anemos.ts:184]: There's just flat out no network packet for this, so I think the "for now" is a bit misleading. The client tracks the idle time locally, separate from the server, and that's what keys this message to fire.

File: ui/raidboss/data/04-sb/dungeon/fractal_continuum_hard.txt [hunk 2]
```diff
@@ -37,7 +37,7 @@ hideall "--sync--"
 
 # -ii 2798
 
-# Unfortunately, we have no choice but to use game log lines here.
+# Unfortunately, we have no choice but to use BattleTalk2 lines here.
 # The statues the boss tethers to are useless for timeline syncs,
 # and the boss gains no buffs for each form.
```
Comment (valarnin) [inline ui/raidboss/data/04-sb/dungeon/fractal_continuum_hard.txt:42]: I'll see if I can check this one to see if it's the case.

File: ui/raidboss/data/04-sb/eureka/eureka_hydatos.ts [hunk 3]
```diff
@@ -77,12 +77,18 @@ const triggerSet: TriggerSet<Data> = {
...
+    // But no 0x29 log line; so have to use GameLog for now.
     {
       id: 'BA Falling Asleep',
       type: 'GameLog',
       netRegex: { line: '7 minutes have elapsed since your last activity..*?', capture: false },
       response: Responses.wakeUp(),
     },
+    // https://xivapi.com/LogMessage/9069
+    // en: The memories of heroes past live on again!
+    // But no 0x29 log line; so have to use GameLog for now.
```
Comment (valarnin) [inline ui/raidboss/data/04-sb/eureka/eureka_hydatos.ts:91]: This one's weird, there's no direct network packet referencing this so I think it's fired as part of some timeline effects or something.

File: ui/raidboss/data/04-sb/eureka/eureka_hydatos.ts [hunk 4]
```diff
@@ -249,6 +255,8 @@ const triggerSet: TriggerSet<Data> = {
       condition: (data, matches) => data.side === 'east' && data.me === matches.target,
       response: Responses.doritoStack(),
     },
+    // https://xivapi.com/PublicContentTextData/2122
+    // But we do not currently have a log line for this data.
```
Comment (valarnin) [inline ui/raidboss/data/04-sb/eureka/eureka_hydatos.ts:259]: I haven't done this one so I can't try and track down the data source.

File: ui/raidboss/data/04-sb/eureka/eureka_hydatos.ts [hunk 5]
```diff
@@ -515,6 +525,7 @@ const triggerSet: TriggerSet<Data> = {
       // In other words, they need to be the fully translated in game log.
       // There are no "gainsEffect" lines for the clones, only for Absolute Virtue directly.
       // Ideally parser logic could be added for this case, but this is where we are.
+      // TODO: Possibly animation data from ActorControlExtra that controls the light/dark wrists?
```
Comment (valarnin) [inline ui/raidboss/data/04-sb/eureka/eureka_hydatos.ts:528]: I think that these effect gains are available now, they were part of the StatusEffectList2 and/or StatusEffectList3 packets which weren't available at the time? But I haven't done the AV fight so I'm not sure.

File: ui/raidboss/data/04-sb/raid/o11n.txt [hunk 6]
```diff
@@ -53,7 +53,11 @@ hideall "--sync--"
 611.6 "Force Quit" Ability { id: "327A", source: "Level Checker" }
 
 ## Delta Attack
-1000.0 "--sync--" GameLog { code: "0044", name: "Omega", line: "Program failure detected.*?" } window 1500,100
+# https://xivapi.com/InstanceContentTextData/20109
+# en: Program failure detected. <blip> Engaging Delta Attack protocol.
+# Note: This BattleTalk2 log line arrives in sequence with another, both with a 4000ms display time.
+# As a result, this line shows up 400ms before the message is displayed (and the GameLog line is emitted).
```
Comment (valarnin) [inline ui/raidboss/data/04-sb/raid/o11n.txt:59]: Is this a typo (`400ms` vs `4000ms`)?

**Comment by valarnin** [comment]
Unrelated to this PR, but it strikes me as a bit odd that there's a BA-specific trigger for Remembrance (and it's typo'd in the trigger name, or in game?), but not for the other Eureka zones...

File: ui/raidboss/data/04-sb/dungeon/fractal_continuum_hard.txt (hunk 2 above)
Comment (JLGarber) [reply ui/raidboss/data/04-sb/dungeon/fractal_continuum_hard.txt:42]: This is my work and it's very likely this isn't true anymore. At the time, the statues were spawned in at the time the boss room was opened, and so there wasn't any reliable way to determine what we were tethered to. I *think* it might be possible to use the tethers in combination with BNPCID or NPCID now, instead of using Battletalk.

File: ui/raidboss/data/04-sb/raid/o12n.txt [hunk 7]
```diff
@@ -181,8 +183,11 @@ hideall "--sync--"
 
 
 ### Phase 4: Laser Show Forever
-1800.0 "Enrage" GameLog { code: "0044", line: "\<blip\> Warning\. Calculations indicate.*?" } window 1800,0
-
+# https://xivapi.com/InstanceContentTextData/20623
+# https://xivapi.com/InstanceContentTextData/20624
+# en: <blip> Warning. Calculations indicate increased probability of defeat.\n\n<bleep> Error. Further data analysis required.
+# May be either id depending on which is alive.
+1800.0 "Enrage" BattleTalk2 { instanceContentTextId: "50(8F|90)" } window 1800,0
```
Comment (JLGarber) [inline ui/raidboss/data/04-sb/raid/o12n.txt:190]: To save ourselves headache if we ever have to do future conversions, let's expand this regex, as `(508F|5090)`.

File: ui/raidboss/data/04-sb/eureka/eureka_anemos.ts (hunk 1 above)
Comment (wexxlee) [reply ui/raidboss/data/04-sb/eureka/eureka_anemos.ts:184]: > There's just flat out no network packet for this, so I think the "for now" is a bit misleading. The client tracks the idle time locally, separate from the server, and that's what keys this message to fire.

Agree & will remove the "for now", but that raises an interesting -- if sort of collateral -- issue.  These GameLog lines use `code` 0039, which [cactbot calls](https://github.com/OverlayPlugin/cactbot/blob/448ecf83b59d287f484054cfbdac71f4849e7abc/resources/netlog_defs.ts#L116) a `message`.

I assumed `code` was coming from the network packet, but if these lines are (or can be) generated by ACT solely from memory reads and not a packet, I think(?) that would mean either ACT or the game client has some reference/table it's using to generate these codes? The LogGuide doesn't have much to say, but quisquous does have a 4+ year old [list](https://gist.github.com/quisquous/1a6ea3cf102c65e3c375186f1173dff3) of what some of these codes may correspond to.

I suppose where I'm going with this is that I'm continuing to run across `GameLog` triggers that don't have a corresponding 0x29 (or other) line, and, it sounds like, may not have an associated network packet at all.  Wondering if it's worth exploring whether OP could do a similar read of chat log lines from memory and write at least some (perhaps based on `type`, and that don't implicate privacy issues) as an OP log line, to avoid the issues of using suppressible `00` lines for triggers/timelines?

File: ui/raidboss/data/04-sb/dungeon/fractal_continuum_hard.txt (hunk 2 above)
Comment (wexxlee) [reply ui/raidboss/data/04-sb/dungeon/fractal_continuum_hard.txt:42]: I'll re-run it as well and see if there are alternatives, but as a general philosophy with this eliminate-GameLog-triggers effort, I'm admittedly trying to preference "replacement" over "rework" where possible, the goal being to reduce the support issues that arise from having chat log hidden for privacy. Since triggering off of something other than the messages (whatever their source) can require adjustments to the timeline, I'm sort of trying to avoid that option unless necessary.

File: ui/raidboss/data/04-sb/eureka/eureka_hydatos.ts (hunk 3 above)
Comment (wexxlee) [reply ui/raidboss/data/04-sb/eureka/eureka_hydatos.ts:91]: > This one's weird, there's no direct network packet referencing this so I think it's fired as part of some timeline effects or something.

That is weird.  IIRC, it would be emitted whenever Spirit of the Remembered procs and revives the player.  The proc chance would have to be server side I would think, so there's probably at least some network packet that signals the game client that the revive has proc'd, even if the actual game log line is being written solely by the client.  I'll see if I can pop into Eureka today/tomorrow and get some more info.

File: ui/raidboss/data/04-sb/eureka/eureka_hydatos.ts (hunk 4 above)
Comment (wexxlee) [reply ui/raidboss/data/04-sb/eureka/eureka_hydatos.ts:259]: > I haven't done this one so I can't try and track down the data source.

Based just on the endpoint (exd) name alone, it *looks* like it would be something close to `BattleTalk2`.  Looking through the derived csv, it looks like nearly all of it is Bozja/Eureka dialog, but some of the earliest (lowest id) rows are actually the dialog used in the Ceremony of Eternal Bonding.  Maybe next time you get a wedding invite...? 😄

File: ui/raidboss/data/04-sb/raid/o11n.txt (hunk 6 above)
Comment (wexxlee) [reply ui/raidboss/data/04-sb/raid/o11n.txt:59]: > Is this a typo (`400ms` vs `4000ms`)?

Oops!

File: ui/raidboss/data/04-sb/eureka/eureka_hydatos.ts (hunk 5 above)
Comment (wexxlee) [reply ui/raidboss/data/04-sb/eureka/eureka_hydatos.ts:528]: > I think that these effect gains are available now, they were part of the StatusEffectList2 and/or StatusEffectList3 packets which weren't available at the time? But I haven't done the AV fight so I'm not sure.

This one was very odd.  I looked through an AV log (thanks @xiashtra !) and it didn't have the `GameLog` lines that these triggers use (and chat log was not hidden).  So I couldn't see how these triggers would have fired, even in current state.

I'll spend a little more time with the log to see if there's an option for using status effects instead.  But testing may be tough to come by - I'm not likely to jump back into BA anytime soon.

**Comment by wexxlee** [comment]
> Unrelated to this PR, but it strikes me as a bit odd that there's a BA-specific trigger for Remembrance (and it's typo'd in the trigger name, or in game?), but not for the other Eureka zones...

My guess is that it's an awareness thing specifically for BA (rather than just general eureka overworld) so a player knows their revive has been burned and to pop another Spirit of the Remembered.

File: ui/raidboss/data/04-sb/eureka/eureka_anemos.ts (hunk 1 above)
Comment (wexxlee) [reply ui/raidboss/data/04-sb/eureka/eureka_anemos.ts]: Interesting!  That seems to check out with some of the other patterns I've seen, e.g., `0039` is used for a fair number of log messages, like "Triple Triad matches not allowed here", but `0839` is used for things like "You change to gunbreaker."  (Although, incidentally, I think `0839` is also the code used for zone seal/unseal messages too.)

Definitely interested in unpacking `code` more, just out of personal curiosity, but in terms of this PR, yeah, I don't think we need a custom log line for idle time, especially since it's the walking definition of a non-combat trigger.  I think we can live with this staying as a `GameLog` trigger for the foreseeable future.

File: ui/raidboss/data/04-sb/eureka/eureka_hydatos.ts (hunk 5 above)
Comment (wexxlee) [reply ui/raidboss/data/04-sb/eureka/eureka_hydatos.ts:528]: > I'll spend a little more time with the log to see if there's an option for using status effects instead. But testing may be tough to come by - I'm not likely to jump back into BA anytime soon.

I'm not seeing any 0x1A lines that could be used as suitable replacements; there are some instances where the boss self-applies `6AE` (Astral Essence) and `6AF` (Umbral Essence), but none of those seem to line up with triple impact stream.  If we were going to replace this trigger, I think we'd need to (a) capture the animation being used by each clone, and then (b) determine the clone firing order.  As for that order, the current logic uses chat lines, but from the log I'm looking at, I don't see those chat lines anywhere, so I can't see how those triggers function now.

We could look at the `StartsUsing` order of the Relative Virtue clones coupled with their animation state... but the position data in this log for the RV actors doesn't look right to me.  The first-firing clone is at (-163,298) and the second is at (-163,330).  It seems strange that both would have the same `x` value, because you're supposed to use their position around the arena (and the light-dark split relative to their position) to figure out which clone is firing next. 
If they are 180 degrees apart, I can't see how you'd figure out which is about to fire, which has me wondering if these RV actors are really the right ones to be looking at.

tldr; this is going to take some runs with logs/vod to unpack, and since I'm unlikely to be back in BA anytime soon, this will likely have to stay an open TODO for now.

**Comment by valarnin** [comment]
Well, while looking for the missing info for some of the GameLog lines in this PR, I found another completely unrelated way that messages are sent...
```
4096|2024-02-15T09:36:36.3310000-05:00|ActorControlSelf|10001234|LogMsg|0000087D|0000001C|00000000|00000000|00000000|00000000|00000000|64cf59b50a89f0bd
00|2024-02-15T09:36:36.0000000-05:00|083C||You cannot carry any more Allagan tomestones of poetics.|afa0fbd62ada5c2f
```
I'll keep digging...

File: ui/raidboss/data/04-sb/eureka/eureka_hydatos.ts (hunk 3 above)
Comment (wexxlee) [reply ui/raidboss/data/04-sb/eureka/eureka_hydatos.ts:91]: Looked at this further, and there's nothing obvious that I can see that would signal a rez attributable to a Spirit of the Remembered proc.  The 0x0C `PlayerStats` line and gained effects like Weakness/Transcendence aren't specific to Remembrance.  So I think this is another one that's going to have to stay as-is.

File: ui/raidboss/data/04-sb/dungeon/fractal_continuum_hard.txt (hunk 2 above)
Comment (wexxlee) [reply ui/raidboss/data/04-sb/dungeon/fractal_continuum_hard.txt:42]: > This is my work and it's very likely this isn't true anymore. At the time, the statues were spawned in at the time the boss room was opened, and so there wasn't any reliable way to determine what we were tethered to. I _think_ it might be possible to use the tethers in combination with BNPCID or NPCID now, instead of using Battletalk.

I looked at it a little more.  The boss does tether to each statue with an 0x23 line, but the tether ids are the same, so all we have to differentiate them on the 0x23 line is combatant ids.  Unfortunately, all of the combatants have the same bNpcNameId and name; the only thing distinguishing them is their x/y position.  Assuming each of the 3 positions always corresponds to the same mode (e.g. Sephirot->Sophia->Zurvan, left to right) -- and I didn't actually run it multiple times to verify that -- we'd have to determine which combatant id corresponds to which primal based on the combatant's position, and then jump/sync based on which combatant id Ultima Warrior tethers to.  That wouldn't be terribly difficult for a set of triggers, but I'm actually not sure how we'd dynamically handle that for a timeline entry.  (Also, the tethers come a bit after the BattleTalk2 packet, so we'd be looking at timeline rework anyway).

Given all that -- and the fact this is content from 2 expacs ago -- I think I'm leaning toward leaving this as BattleTalk2

File: ui/raidboss/data/04-sb/eureka/eureka_hydatos.ts (hunk 5 above)
Comment (xiashtra) [reply ui/raidboss/data/04-sb/eureka/eureka_hydatos.ts:528]: > > I think that these effect gains are available now, they were part of the StatusEffectList2 and/or StatusEffectList3 packets which weren't available at the time? But I haven't done the AV fight so I'm not sure.
> 
> This one was very odd. I looked through an AV log (thanks @xiashtra !) and it didn't have the `GameLog` lines that these triggers use (and chat log was not hidden). So I couldn't see how these triggers would have fired, even in current state.
> 
> I'll spend a little more time with the log to see if there's an option for using status effects instead. But testing may be tough to come by - I'm not likely to jump back into BA anytime soon.

I looked through some other BA logs I have saved and found a few that do have the lines from these triggers, but they are only in ~1/3 of the BA logs I have and they are not from a consistent time period.  I'm going to try and get into a few BA runs and see if I can investigate more why they are not being written consistently.

**Comment by xiashtra** [comment]
> > Unrelated to this PR, but it strikes me as a bit odd that there's a BA-specific trigger for Remembrance (and it's typo'd in the trigger name, or in game?), but not for the other Eureka zones...
> 
> My guess is that it's an awareness thing specifically for BA (rather than just general eureka overworld) so a player knows they're now alive and back in combat (probably), since it can be an easy thing to miss if you're not paying attention.

I don't know who wrote the original BA triggers, but I imagine it was because an auto-raise from Spirit of the Remembered (or Resistance Reraiser/Lost Reraise in Bozja zones) happens immediately without user interaction, unlike normal resurrections where you can choose when to accept the raise.

File: ui/raidboss/data/04-sb/eureka/eureka_hydatos.ts (hunk 5 above)
Comment (xiashtra) [reply ui/raidboss/data/04-sb/eureka/eureka_hydatos.ts:528]: Per Discord conversation with @wexxlee, it seems that the regex on these triggers isn't correct; they aren't firing due to special characters present in the chatline:
```00|2024-03-17T22:44:49.0000000-04:00|332E||Relative Virtue gains the effect of Umbral Essence.|98acce60df5f0a61
00|2024-03-17T22:44:49.0000000-04:00|332E||Relative Virtue gains the effect of Umbral Essence.|cd0cc61b9cc9881f
00|2024-03-17T22:44:50.0000000-04:00|332E||Relative Virtue gains the effect of Astral Essence.|48e5f0b52de318d1```
Furthermore, it appears that these chatlines aren't enabled by the default XIV Log Filters; I had to enable the options to show `Beneficial effects granted by unengaged enemies.` and `Beneficial effects granted to unengaged enemies.` to have them written to the log in the first place.

Comment (wexxlee) [reply ui/raidboss/data/04-sb/eureka/eureka_hydatos.ts:528]: I'll clean up the regex to account for the unicode, but if these log lines are buried behind a game setting (that's off by default), we *really* should replace these triggers.

Disregard my earlier comment about the position data being wonky.  The clones were on intercards, hence the same x value (because, yeah, that's how circles work).  In any case, we can determine clone firing order based on the order of each Relative Virtue combatant's 0x14 line for `3795` (Impact Stream).

What we don't have a solve for right now is determining their effect/animation to indicate which color is safe.  There's no 0x1A, 0x26, or 0x2A lines for the `6AE` Astral Essence/`6AF` Umbral Essence effects.

@valarnin - since we've confirmed these chat lines are (or at least can be) written, is it possible there's a `StatusEffectList` packet, but no line emitted from ACT because the combatant is invisible/unengaged when the effect is being applied?  If so, maybe this is justification for a `StatusListExtra` or similar OP line that would provide status effect data for these combatants (assuming of course that is the issue)?

Comment (valarnin) [reply ui/raidboss/data/04-sb/eureka/eureka_hydatos.ts]: If someone's got a log with packet capture enabled, I could check. Otherwise I'll have to get back in to Eureka and run it myself to check.

I don't even know where AV is tbh, I did a single run of Eureka and never saw it so I'm guessing maybe it's in the support group side or something?

Comment (Akurosia) [reply ui/raidboss/data/04-sb/eureka/eureka_hydatos.ts]: > If someone's got a log with packet capture enabled, I could check. Otherwise I'll have to get back in to Eureka and run it myself to check.
> 
> I don't even know where AV is tbh, I did a single run of BA and never saw it so I'm guessing maybe it's in the support group side or something?

AV is the 3rd Boss inside BA just before the Final Boss staircase (where you have to wait for the support group to finish the special fate outside before entering the fight to weaken AV)

Comment (valarnin) [reply ui/raidboss/data/04-sb/eureka/eureka_hydatos.ts]: Hmm, maybe the day I ran BA was the day my ACT crashed and I didn't notice... I'll have to get another run in to check I guess.

Comment (wexxlee) [reply ui/raidboss/data/04-sb/eureka/eureka_hydatos.ts]: @xiashtra, if you're headed back in again sometime soon, think you could enable packet capture for AV?

Comment (valarnin) [reply ui/raidboss/data/04-sb/eureka/eureka_hydatos.ts]:
```
26|2024-03-18T19:11:44.1650000-04:00|6AF|Umbral Essence|9999.00|400027B0|Absolute Virtue|400027B0|Absolute Virtue|00|5858587|5858587|678ca4361787d1c2
```
???

I'm seeing the status effect in opcode `641` (0x281) for these actors (not posting the full lines since it has player IDs in it, in reverse byte order, so they're annoying to filter out), timestamped `2024-03-18T19:11:44.1740000-04:00`:
```
40002650
40002652
40002653
40002651
40002654
40002655
40002657
40002659
40002656
40002658
4000265A
```

But the chat line for this happens before either of the actual network packets (raw packet or the `26` line) get dumped to file? And opcode `641` doesn't line up with any of the opcodes in <https://github.com/ravahn/machina/blob/master/Machina.FFXIV/Headers/Opcodes/Global.txt>, I would have expected that to be the opcode for `StatusEffectList3`...

Looking even further into this, I think that the status effect is actually encoded on the `21` line for the `Eidos` cast itself here, in the `Damage` field?
```
21|2024-03-18T19:11:43.5460000-04:00|400027B0|Absolute Virtue|3787|Eidos|400027B0|Absolute Virtue|E|6AF0000|1B|37878000|0|0|0|0|0|0|0|0|0|0|0|0|4699508|5858587|10000|10000|||-189.90|313.80|82.00|-1.56|4699508|5858587|10000|10000|||-189.90|313.80|82.00|-1.56|0001AA58|0|1|ad8433ddd5799e0e
flags = 0xE
damage high = 0x06AF
damage low = 0x0000
```

Later on with the 3 clones, it looks like they only gain the buff via ActorControl packets, so ACT never emits a normal log line for those? (this was a different run, ignore the large difference in timestamp to previous examples)

```
4096|2024-03-18T20:54:39.6610000-04:00|ActorControl|40005B89|GainEffect|000006AF|00000000|00000000|00000000|00000000|05e43a2eaa7cb1c5
4096|2024-03-18T20:54:39.6610000-04:00|ActorControl|40005B8C|GainEffect|000006AE|00000000|00000000|00000000|00000000|3fb5d121f87c880a
4096|2024-03-18T20:54:39.6610000-04:00|ActorControl|40005B8F|GainEffect|000006AE|00000000|00000000|00000000|00000000|755704f2396d4ea0

00|2024-03-18T20:54:39.0000000-04:00|332E||Relative Virtue gains the effect of Umbral Essence.|1d1901a5943f9f31
00|2024-03-18T20:54:39.0000000-04:00|332E||Relative Virtue gains the effect of Astral Essence.|28053d2aebea695a
00|2024-03-18T20:54:39.0000000-04:00|332E||Relative Virtue gains the effect of Astral Essence.|b0cdcaa506042cca
```

Unfortunately all 6 instances of `Eidos` I have from today were Umbral Essence so I can't confirm that Astral will have the correct ID in the `Eidos` cast.

File: ui/raidboss/data/04-sb/eureka/eureka_hydatos.ts (hunk 3 above)
Comment (valarnin) [reply ui/raidboss/data/04-sb/eureka/eureka_hydatos.ts:91]: Well I finally found the network packet indicating this one, it's ActorControlSelf category 527 (0x20F):
```
4096|2024-03-18T12:34:52.2740000-04:00|ActorControlSelf|10001234|527|0000236D|00000000|00000669|00000000|00000000|00000000|00000000|450d8d6ada5c140f
00|2024-03-18T12:34:51.0000000-04:00|0939||The memories of heroes past live on again!|e710cf321339b40d
```

So another potential item to add to the ActorControl packet handler.

File: ui/raidboss/data/04-sb/eureka/eureka_hydatos.ts (hunk 4 above)
Comment (valarnin) [reply ui/raidboss/data/04-sb/eureka/eureka_hydatos.ts:259]: This looks to also be an `ActorControl` packet, I think. Still haven't done that side of BA yet, but an equivalent line from Art:
```
4096|2024-03-18T20:33:22.7130000-04:00|ActorControl|400058CA|2100|00000000|00000848|00000FA0|00000000|00000000|c862c35712ed4122
00|2024-03-18T20:33:22.0000000-04:00|0044|Art|O spear unyielding, become thee a thousand soldiers!|cf0c282daa64b90b
```

`0x848` = <https://xivapi.com/PublicContentTextData/2120>

File: ui/raidboss/data/04-sb/eureka/eureka_hydatos.ts (hunk 5 above)
Comment (wexxlee) [reply ui/raidboss/data/04-sb/eureka/eureka_hydatos.ts]: > ???
> 
> I'm seeing the status effect in opcode `641` (0x281) for these actors (not posting the full lines since it has player IDs in it, in reverse byte order, so they're annoying to filter out), timestamped `2024-03-18T19:11:44.1740000-04:00`:
> [...]
> 
> Looking even further into this, I think that the status effect is actually encoded on the `21` line for the `Eidos` cast itself here, in the `Damage` field?

I might not be fully tracking your thoughts here, but I don't think the status effect on Absolute Virtue (or AV hidden combatants) is necessarily the data point needed to resolve the mechanic?  There is at least one (maybe multiple? can't recall) instances where the main boss does Eidos and then a single Impact Stream, and I think there are hidden AV combatants to handle the vfx/damage.

From looking at the data for Triple Impact Stream, though, it's the 3 Relative Virtue Clones with the glowing wrists (`6AE`/`6AF`) that determine which half of the arena is unsafe.  At the same position as each RV actor, there are also two hidden AV actors, each facing 90 degrees away from the RV actor's heading (which should be toward the middle of the arena), each one doing either Bright Aurora or Dark Aurora on the half they are facing.  But knowing that (or which buff they have) wouldn't, I don't think, tell us which half will be safe?  (Sorry if I'm misunderstanding your point.)

> Later on with the 3 clones, it looks like they only gain the buff via ActorControl packets, so ACT never emits a normal log line for those? (this was a different run, ignore the large difference in timestamp to previous examples)
> 
> ```
> 4096|2024-03-18T20:54:39.6610000-04:00|ActorControl|40005B89|GainEffect|000006AF|00000000|00000000|00000000|00000000|05e43a2eaa7cb1c5
> 4096|2024-03-18T20:54:39.6610000-04:00|ActorControl|40005B8C|GainEffect|000006AE|00000000|00000000|00000000|00000000|3fb5d121f87c880a
> 4096|2024-03-18T20:54:39.6610000-04:00|ActorControl|40005B8F|GainEffect|000006AE|00000000|00000000|00000000|00000000|755704f2396d4ea0
> 
> 00|2024-03-18T20:54:39.0000000-04:00|332E||Relative Virtue gains the effect of Umbral Essence.|1d1901a5943f9f31
> 00|2024-03-18T20:54:39.0000000-04:00|332E||Relative Virtue gains the effect of Astral Essence.|28053d2aebea695a
> 00|2024-03-18T20:54:39.0000000-04:00|332E||Relative Virtue gains the effect of Astral Essence.|b0cdcaa506042cca
> ```

Yep, I think those are the packets we need.  Since each one of those hidden actors will then cast `Impact Stream` in sequence (cast time of 4.7 seconds, with a little wiggle room in between for damage application), if we've pre-collected this packet data, we should be able to call out whether Light or Dark is safe for each cast.

Looking at the ActorControl volume table you put together, `GainEffect` is up there, relatively speaking, in terms of "noisy" categories.  And this seems like a fairly niche case, so reasonable minds could differ on whether it's worth it to add it to `ActorControlExtra`.  But I don't think there's any other way to resolve this mechanic with the data we have, so I'd be in favor of adding it.  (I suppose there's always the option to add this category just for this particular zone, but .... yeah, not sure how I feel about OP filtering the lines it emits based on zone.)

File: ui/raidboss/data/04-sb/eureka/eureka_hydatos.ts (hunk 4 above)
Comment (wexxlee) [reply ui/raidboss/data/04-sb/eureka/eureka_hydatos.ts:259]: > This looks to also be an `ActorControl` packet, I think. Still haven't done that side of BA yet, but an equivalent line from Art:
> 
> ```
> 4096|2024-03-18T20:33:22.7130000-04:00|ActorControl|400058CA|2100|00000000|00000848|00000FA0|00000000|00000000|c862c35712ed4122
> 00|2024-03-18T20:33:22.0000000-04:00|0044|Art|O spear unyielding, become thee a thousand soldiers!|cf0c282daa64b90b
> ```
> 
> `0x848` = https://xivapi.com/PublicContentTextData/2120

It looks like `2100` didn't even make the list of `ActorControl` packet volume - which is unsurprising considering that `PublicContentTextData` seems to be largely (if not exclusively) limited to Eureka, Bozja/Zadnor, and weddings.  I'd be in favor of adding this to `ActorControlExtra` as well.  (Purely supposition, but this could also be some future-proofing for Cosmic Exploration?)

File: ui/raidboss/data/04-sb/eureka/eureka_hydatos.ts (hunk 3 above)
Comment (wexxlee) [reply ui/raidboss/data/04-sb/eureka/eureka_hydatos.ts:91]: > Well I finally found the network packet indicating this one, it's ActorControlSelf category 527 (0x20F):
> 
> ```
> 4096|2024-03-18T12:34:52.2740000-04:00|ActorControlSelf|10001234|527|0000236D|00000000|00000669|00000000|00000000|00000000|00000000|450d8d6ada5c140f
> 00|2024-03-18T12:34:51.0000000-04:00|0939||The memories of heroes past live on again!|e710cf321339b40d
> ```
> 
> So another potential item to add to the ActorControl packet handler.

Works for me.

**Review by valarnin — APPROVED** [review]
I think that some `TODO` comments above the two open comment locations above referencing back to this PR should be fine for now, then this can be merged and those can be cleaned up in a followup PR once the relevant issues have been fixed upstream.

File: ui/raidboss/data/04-sb/eureka/eureka_hydatos.ts (hunk 5 above)
Comment (wexxlee) [reply ui/raidboss/data/04-sb/eureka/eureka_hydatos.ts]: > I have an open question to Ravahn in the dev channel of the discord about the `Relative Virtue` case.

I'm planning to merge with a todo/comment - but assuming we don't end up with a definitive answer, do you have thoughts on whether it would be worthwhile to add this to `ActorControlExtra` (and if/how we should filter to prevent duplication of existing 0x1A lines?  Perhaps just these two effects?)
